src/io/streams.rs

```diff
@@ -40,7 +40,12 @@ impl AsyncRead for AsyncInputStream {
         // Ideally, the ABI would be able to read directly into buf. However, with the default
         // generated bindings, it returns a newly allocated vec, which we need to copy into buf.
         let read = match self.stream.read(buf.len() as u64) {
+            // 0 bytets from WASI's `read` just means we got zero bytes.
+            Ok(r) if r.is_empty() => {
+                return Err(std::io::Error::from(std::io::ErrorKind::Interrupted))
```

Comment:
According to the wit spec, an empty read after ready can only indicate that the stream is still open iff buf.len() is zero - otherwise, read must return something (https://github.com/WebAssembly/wasi-io/blob/main/wit/streams.wit#L53). Rust's Read::read documents the same - its valid to return Ok(0) if buf.len() ==0, or the stream has been closed (taken care of by next case) https://doc.rust-lang.org/std/io/trait.Read.html#tymethod.read

Reply:
Ah, I was just reading the "The list is empty when no bytes are available for reading at this time." part and missed the implications of the `self.ready()` call. I've now removed that change and added a comment.

Reply:
Yeah I should probably try to rewrite that doc to make that more clear...